### PR TITLE
Issue #7634: Make `UndocumentedPublicClass` configurable to flag `com…

### LINKED ISSUE #7634: Regression: 1.23.7 - `UndocumentedPublicClass` flags unnamed public companion object
## Expected Behavior
`UndocumentedPublicClass` is _configurable_ to flag unnamed `companion object`. 

## Observed Behavior
`UndocumentedPublicClass` flags `companion object` newly in 1.23.7

## Steps to Reproduce
```kotlin
/** Some doc **/
public class Foo {
  public companion object // Flagged
}
```

## Context

In #7219, this behaviour was enabled. The rule is written/described such that this is expected. However, this was a breaking change as it caused significant amount of violations within our code base. 

## Your Environment
* Version of detekt used: 1.23.7
* Version of Gradle used (if applicable): 8.10
* Operating System and version: macos

### PR DESCRIPTION
**Pull request by Nava2**
…panion object` without name

Closes #7634 in favour of leaving the rule as described with configuration to disable to pre-#7219 value.

### CODE REVIEW

**Comment by Nava2** [comment]
@3flex / @cortinico I see you're regularly reviewing PRs, is there a step I'm expected to take to flag changes to maintainers? I checked the documentation but found nothing 😢. This small configuration fix would significantly help our personal migration.

File: detekt-rules-documentation/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/UndocumentedPublicClass.kt [hunk 1]
```diff
@@ -41,6 +41,9 @@ class UndocumentedPublicClass(config: Config) : Rule(
     @Configuration("if protected classes should be searched")
     private val searchInProtectedClass: Boolean by config(false)
 
+    @Configuration("if companion object without names should be flagged")
```
Comment (Nava2) [inline detekt-rules-documentation/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/UndocumentedPublicClass.kt:44]:
```suggestion
    @Configuration("if companion object without a name should be flagged")
```

**Review by BraisGabin — APPROVED** [review]

**Comment by BraisGabin** [comment]
LGTP. I would like to know what other maintainers think about this. @TWiStErRob you did the initial request to enable this, what do you thik about this?

File: detekt-rules-documentation/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/UndocumentedPublicClassSpec.kt [hunk 2]
```diff
@@ -269,6 +270,21 @@ class UndocumentedPublicClassSpec {
         assertThat(subject.compileAndLint(code)).hasSize(3)
     }
 
+    @Test
+    fun `should not report in public companion class if disabled`() {
+        val code = """
+            /** Some doc */
+            public class PublicClass {
+                public companion object
+            }
```
Comment (TWiStErRob) [inline detekt-rules-documentation/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/UndocumentedPublicClassSpec.kt:279]: I wonder if the existence of any content matters for this feature?

```suggestion
            public class PublicClass {
                public companion object {
                    public val x: String = ""
                }
            }
```

File: detekt-core/src/main/resources/default-detekt-config.yml [hunk 3]
```diff
@@ -71,6 +71,7 @@ comments:
     searchInInnerObject: true
     searchInInnerInterface: true
     searchInProtectedClass: false
+    flagCompanionWithoutName: true
```
Comment (TWiStErRob) [inline detekt-core/src/main/resources/default-detekt-config.yml:74]: [docs](https://kotlinlang.org/docs/object-declarations.html#companion-objects) refer to this as "unnamed companion object", so I would propose the name:
```suggestion
    ignoreUnnamedCompanionObject: false
```
which
1. matches the official terminology
2. matches params style of other rules
3. somehow makes the logic simpler (not `!`)
   _Note: normally I would advocate for a positively phrased configuration entry (`allowUnnamedCompanionObject`), but somehow in this case `ignore` makes sense as an "opt in" to more lenient behavior._

**Review by TWiStErRob — COMMENTED** [review]
Make sense to have this feature, two thoughts.

File: detekt-rules-documentation/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/UndocumentedPublicClassSpec.kt (hunk 2 above)
Comment (Nava2) [reply detekt-rules-documentation/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/UndocumentedPublicClassSpec.kt:296]: I duplicated the test, one for with and one for without content!

File: detekt-core/src/main/resources/default-detekt-config.yml (hunk 3 above)
Comment (Nava2) [reply detekt-core/src/main/resources/default-detekt-config.yml]: I agree with all of these points, I'm even a bit surprised to agree with the negative phrasing, I actually had it this way before but changed it due to positive phrasing 👍🏻

**Comment by Nava2** [comment]
Given how impactful this change was in existing codebases, I wonder if this should be considered for 1.23.8 rather than 2.0.0.

File: detekt-rules-documentation/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/UndocumentedPublicClass.kt [hunk 4]
```diff
@@ -41,6 +41,9 @@ class UndocumentedPublicClass(config: Config) : Rule(
     @Configuration("if protected classes should be searched")
     private val searchInProtectedClass: Boolean by config(false)
 
+    @Configuration("if companion object without a name should be flagged")
```
Comment (TWiStErRob) [inline detekt-rules-documentation/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/UndocumentedPublicClass.kt:44]: Outdated doc?

File: detekt-rules-documentation/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/UndocumentedPublicClassSpec.kt (hunk 2 above)
Comment (TWiStErRob) [reply detekt-rules-documentation/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/UndocumentedPublicClassSpec.kt:296]: I didn't mean we need more tests (although more edge case coverage never hurts), but whether it changes the "ignore" logic in any way if there's content vs just an empty placeholder for extensions? Do your use cases apply to all unnamed?

File: detekt-rules-documentation/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/UndocumentedPublicClass.kt (hunk 4 above)
Comment (Nava2) [reply detekt-rules-documentation/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/UndocumentedPublicClass.kt:44]:
```suggestion
    @Configuration("if an unnamed companion object should be allowed")
```

File: detekt-rules-documentation/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/UndocumentedPublicClassSpec.kt (hunk 2 above)
Comment (Nava2) [reply detekt-rules-documentation/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/UndocumentedPublicClassSpec.kt:296]: What are you referring to by "all unnamed?" Apologies, I'm not quite following 😅

Comment (TWiStErRob) [reply detekt-rules-documentation/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/UndocumentedPublicClassSpec.kt:296]: all unnamed including body or not.

I see 4 categories of companion objects (syntactically):
* unnamed marker (`companion object;`) - for extensions
* unnamed default (`companion object {}`) - with val's and funs inside = "static" stuff
* named default (`companion object Companion ...`) - not sure why, but this feels like an edge case
* named (`companion object Foo`)

So my thinking is that not all of these have the same rules. For example a named companion object (4th bullet) should definitely always have docs (at minimum to explain the name change). And my original question was whether 1st and 2nd have the same rule regarding docs and the exemption you're adding.

File: detekt-rules-documentation/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/UndocumentedPublicClass.kt (hunk 4 above)
Comment (TWiStErRob) [reply detekt-rules-documentation/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/UndocumentedPublicClass.kt:44]: I would go with 
```suggestion
    @Configuration("whether unnamed companion objects should be exempted from this rule")
```

File: detekt-rules-documentation/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/UndocumentedPublicClassSpec.kt (hunk 2 above)
Comment (Nava2) [reply detekt-rules-documentation/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/UndocumentedPublicClassSpec.kt:296]: > all unnamed including body or not.
> 
> I see 4 categories of companion objects (syntactically):
> * unnamed marker (`companion object;`) - for extensions
> * unnamed default (`companion object {}`) - with val's and funs inside = "static" stuff

I'd argue these are the same class of usages. I think both should be covered under the "unnamed" exception as currently defined. I don't think there's much of a difference in usages. Anecdotally, I often see both usages mixing over time. What may start as an extension only usage, can morph into some funs and vals!

> * named default (`companion object Companion ...`) - not sure why, but this feels like an edge case

This definitely feels like an edge case 😅 

> * named (`companion object Foo`)

This would not be covered by the exemption (hence unnamed!), and it feels like we can let the edge case fall into this one. 

All in, I think we've covered our cases! But let me know if you think differently. :+1:

Comment (TWiStErRob) [reply detekt-rules-documentation/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/UndocumentedPublicClassSpec.kt:296]: Makes sense! Thanks for clarifying questions and thoughts.

**Comment by Nava2** [comment]
@TWiStErRob added a couple more tests and fixed a bug with that edge case. LMK your thoughts!

File: detekt-rules-documentation/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/UndocumentedPublicClass.kt [hunk 5]
```diff
@@ -92,8 +93,12 @@ class UndocumentedPublicClass(config: Config) : Rule(
...
             )
 
-    private fun KtObjectDeclaration.isCompanionWithoutName() =
-        isCompanion() && nameAsSafeName.asString() == "Companion"
+    private fun KtObjectDeclaration.isDefaultCompanionObject() =
+        isCompanion() &&
+            nameAsSafeName.asString() == "Companion" &&
+            // For companions _named_ `Companion`, we want to check if the name is a duplicate occurance. If so, it's
+            // not a companion without name.
+            text.lowercase().countOccurrencesOf("companion") == 1
```
[the remark this answers is not in the document]
Comment (Nava2) [reply detekt-rules-documentation/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/UndocumentedPublicClass.kt:101]: shoot, I think you're right. Wonder I can narrow to the declaration. The fun part is that the `default` name ends up being `Companion`, but it can also end up being `public` or `object` because there seems to be a bug where we change the `name` to be the "first" element if it is defined as `Companion`. 😢 

Let me see if I can make this more targetted.

[the remark this answers is not in the document]
Comment (Nava2) [reply detekt-rules-documentation/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/UndocumentedPublicClass.kt]: It's definitely more targetted now, used `nameIdentifier?.text == "Companion"` which is _only_ true in the edge case. Thanks for the push!

**Comment by cortinico** [comment]
> Given how impactful this change was in existing codebases, I wonder if this should be considered for 1.23.8 rather than 2.0.0.

Can you elaborate on this? This feels like a new feature and not a bugfix to me

**Comment by Nava2** [comment]
> > Given how impactful this change was in existing codebases, I wonder if this should be considered for 1.23.8 rather than 2.0.0.
> 
> Can you elaborate on this? This feels like a new feature and not a bugfix to me

Sure, upgrading to the most recent 1.2x release caused 100k+ errors on our codebase on an *existing* rule. If this was a _newly added_ rule, I'd agree it was a feature change. IMO the previous change to add this functionality was somewhere between a regression and a feature given the impact. The default companion are so widespread it's difficult to make "non-breaking" changes.

**Comment by cortinico** [comment]
> Sure, upgrading to the most recent 1.2x release caused 100k+ errors on our codebase on an _existing_ rule

Ah I see. This was introduced by this pick:
- https://github.com/detekt/detekt/pull/7219

So yes we can ship a 1.x release with this fix included 👍

**Comment by Nava2** [comment]
> > Sure, upgrading to the most recent 1.2x release caused 100k+ errors on our codebase on an _existing_ rule
> 
> Ah I see. This was introduced by this pick:
> 
>     * [Check for `public companion` object for `UndocumentedPublicClass` #7219](https://github.com/detekt/detekt/pull/7219)
> 
> 
> So yes we can ship a 1.x release with this fix included 👍

Great! Thanks! I think we're just waiting on a stamp from @TWiStErRob, then. 👍🏻

**Comment by Nava2** [comment]
@TWiStErRob Bump here :)

**Review by TWiStErRob — APPROVED** [review]
Stamp

**Comment by zippert** [comment]
Hello. Any plans to release a new version with this fix anytime soon?

File: detekt-rules-documentation/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/UndocumentedPublicClass.kt [hunk 6]
```diff
@@ -41,6 +41,9 @@ class UndocumentedPublicClass(config: Config) : Rule(
     @Configuration("if protected classes should be searched")
     private val searchInProtectedClass: Boolean by config(false)
 
+    @Configuration("whether default companion objects should be exempted")
+    private val ignoreDefaultCompanionObject: Boolean by config(false)
```
Comment (eboudrant) [inline detekt-rules-documentation/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/UndocumentedPublicClass.kt:45]: Nice, I was looking for config for the same issue. I think documentation need to be updated?

https://detekt.dev/docs/rules/comments/#undocumentedpublicclass

Comment (TWiStErRob) [reply detekt-rules-documentation/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/UndocumentedPublicClass.kt:45]: Change the version to "next" (2.x):
image

https://detekt.dev/docs/next/rules/comments/#undocumentedpublicclass

Comment (eboudrant) [reply detekt-rules-documentation/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/UndocumentedPublicClass.kt:45]: oops, thanks and sorry!
